File: packages/java-sdk/src/test/java/com/turbodocx/TurboPartnerWebhooksTest.java
package com.turbodocx;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * TurboPartner Webhook Management Tests
 *
 * Tests for webhook provisioning operations:
 * - createWebhook
 * - listWebhooks
 * - getWebhook
 * - updateWebhook
 * - deleteWebhook
 * - testWebhook
 * - listWebhookDeliveries
 */
class TurboPartnerWebhooksTest {

    private MockWebServer server;
    private TurboPartnerClient client;
    private final Gson gson = new Gson();

    private static final String PARTNER_ID = "test-partner-id";
    private static final String ORG_ID = "org-uuid-456";
    private static final String WEBHOOK_NAME = "my-signing-webhook";

    @BeforeEach
    void setUp() throws IOException {
        server = new MockWebServer();
        server.start();

        client = new TurboPartnerClient.Builder()
                .partnerApiKey("TDXP-test-key")
                .partnerId(PARTNER_ID)
                .baseUrl(server.url("/").toString())
                .build();
    }

    @AfterEach
    void tearDown() throws IOException {
        server.shutdown();
    }

    // ============================================
    // createWebhook
    // ============================================

    @Test
    @DisplayName("should create a webhook and return result with secret")
    void createWebhookSuccess() throws Exception {
        JsonObject responseData = new JsonObject();
        responseData.addProperty("id", "webhook-uuid-789");
        responseData.addProperty("name", WEBHOOK_NAME);
        responseData.addProperty("isActive", true);
        responseData.addProperty("secret", "whsec_abc123");

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().createWebhook(
                ORG_ID,
                WEBHOOK_NAME,
                List.of("https://example.com/hook"),
                List.of("signature.document.completed"),
                null
        );

        RecordedRequest request = server.takeRequest();
        assertEquals("POST", request.getMethod());
        assertEquals("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks", request.getPath());
        assertEquals("Bearer TDXP-test-key", request.getHeader("Authorization"));

        assertTrue((Boolean) result.get("success"));
        @SuppressWarnings("unchecked")
        Map<String, Object> data = (Map<String, Object>) result.get("data");
        assertEquals(WEBHOOK_NAME, data.get("name"));
        assertEquals("whsec_abc123", data.get("secret"));
    }

    @Test
    @DisplayName("should throw AuthenticationException on 401")
    void createWebhookUnauthorized() {
        server.enqueue(new MockResponse()
                .setResponseCode(401)
                .setHeader("Content-Type", "application/json")
                .setBody("{\"error\":\"Unauthorized\"}"));

        assertThrows(TurboDocxException.AuthenticationException.class, () ->
                client.turboPartner().createWebhook(
                        ORG_ID, WEBHOOK_NAME,
                        List.of("https://example.com/hook"),
                        List.of("signature.document.completed"),
                        null
                ));
    }

    // ============================================
    // listWebhooks
    // ============================================

    @Test
    @DisplayName("should list webhooks with pagination")
    void listWebhooksWithPagination() throws Exception {
        JsonObject responseData = new JsonObject();
        responseData.addProperty("totalRecords", 1);
        responseData.addProperty("limit", 10);
        responseData.addProperty("offset", 0);
        responseData.add("results", gson.toJsonTree(List.of(
                Map.of("id", "wh-1", "name", WEBHOOK_NAME)
        )));

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().listWebhooks(ORG_ID, 10, 0, null);

        RecordedRequest request = server.takeRequest();
        assertEquals("GET", request.getMethod());
        assertTrue(request.getPath().startsWith("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks"));
        assertTrue(request.getPath().contains("limit=10"));

        assertTrue((Boolean) result.get("success"));
    }

    @Test
    @DisplayName("should list webhooks without params")
    void listWebhooksNoParams() throws Exception {
        JsonObject responseData = new JsonObject();
        responseData.addProperty("totalRecords", 0);
        responseData.addProperty("limit", 50);
        responseData.addProperty("offset", 0);
        responseData.add("results", gson.toJsonTree(List.of()));

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().listWebhooks(ORG_ID, null, null, null);

        assertTrue((Boolean) result.get("success"));
    }

    // ============================================
    // getWebhook
    // ============================================

    @Test
    @DisplayName("should get webhook by name")
    void getWebhookByName() throws Exception {
        JsonObject responseData = new JsonObject();
        responseData.addProperty("id", "webhook-uuid-789");
        responseData.addProperty("name", WEBHOOK_NAME);

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().getWebhook(ORG_ID, WEBHOOK_NAME);

        RecordedRequest request = server.takeRequest();
        assertEquals("GET", request.getMethod());
        assertEquals("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks/" + WEBHOOK_NAME, request.getPath());

        assertTrue((Boolean) result.get("success"));
    }

    @Test
    @DisplayName("should throw NotFoundException on 404")
    void getWebhookNotFound() {
        server.enqueue(new MockResponse()
                .setResponseCode(404)
                .setHeader("Content-Type", "application/json")
                .setBody("{\"error\":\"Not Found\"}"));

        assertThrows(TurboDocxException.NotFoundException.class, () ->
                client.turboPartner().getWebhook(ORG_ID, "nonexistent"));
    }

    // ============================================
    // updateWebhook
    // ============================================

    @Test
    @DisplayName("should update webhook isActive field")
    void updateWebhookIsActive() throws Exception {
        JsonObject responseData = new JsonObject();
        responseData.addProperty("id", "webhook-uuid-789");
        responseData.addProperty("name", WEBHOOK_NAME);
        responseData.addProperty("isActive", false);

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> updates = new HashMap<>();
        updates.put("isActive", false);

        Map<String, Object> result = client.turboPartner().updateWebhook(ORG_ID, WEBHOOK_NAME, updates);

        RecordedRequest request = server.takeRequest();
        assertEquals("PATCH", request.getMethod());
        assertEquals("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks/" + WEBHOOK_NAME, request.getPath());

        assertTrue((Boolean) result.get("success"));
    }

    // ============================================
    // deleteWebhook
    // ============================================

    @Test
    @DisplayName("should delete webhook successfully")
    void deleteWebhookSuccess() throws Exception {
        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.addProperty("message", "Webhook deleted");

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().deleteWebhook(ORG_ID, WEBHOOK_NAME);

        RecordedRequest request = server.takeRequest();
        assertEquals("DELETE", request.getMethod());
        assertEquals("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks/" + WEBHOOK_NAME, request.getPath());

        assertTrue((Boolean) result.get("success"));
    }

    // ============================================
    // testWebhook
    // ============================================

    @Test
    @DisplayName("should send test event and return delivery summary")
    void testWebhookDefault() throws Exception {
        JsonObject summary = new JsonObject();
        summary.addProperty("total", 1);
        summary.addProperty("successful", 1);
        summary.addProperty("failed", 0);
        summary.add("errors", gson.toJsonTree(List.of()));

        JsonObject responseData = new JsonObject();
        responseData.add("deliveries", gson.toJsonTree(List.of()));
        responseData.add("summary", summary);

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().testWebhook(ORG_ID, WEBHOOK_NAME, null);

        RecordedRequest request = server.takeRequest();
        assertEquals("POST", request.getMethod());
        assertEquals("/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks/" + WEBHOOK_NAME + "/test", request.getPath());

        assertTrue((Boolean) result.get("success"));
    }

    @Test
    @DisplayName("should pass event override to test endpoint")
    void testWebhookWithEventOverride() throws Exception {
        JsonObject summary = new JsonObject();
        summary.addProperty("total", 1);
        summary.addProperty("successful", 1);
        summary.addProperty("failed", 0);
        summary.add("errors", gson.toJsonTree(List.of()));

        JsonObject responseData = new JsonObject();
        responseData.add("deliveries", gson.toJsonTree(List.of()));
        responseData.add("summary", summary);

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> overrides = new HashMap<>();
        overrides.put("event", "signature.document.voided");

        Map<String, Object> result = client.turboPartner().testWebhook(ORG_ID, WEBHOOK_NAME, overrides);

        RecordedRequest request = server.takeRequest();
        String body = request.getBody().readUtf8();
        assertTrue(body.contains("signature.document.voided"));

        assertTrue((Boolean) result.get("success"));
    }

    // ============================================
    // listWebhookDeliveries
    // ============================================

    @Test
    @DisplayName("should list webhook deliveries with pagination")
    void listWebhookDeliveries() throws Exception {
        JsonObject delivery = new JsonObject();
        delivery.addProperty("id", "del-1");
        delivery.addProperty("event", "signature.document.completed");
        delivery.addProperty("statusCode", 200);

        JsonObject responseData = new JsonObject();
        responseData.addProperty("totalRecords", 1);
        responseData.addProperty("limit", 50);
        responseData.addProperty("offset", 0);
        responseData.add("results", gson.toJsonTree(List.of(gson.fromJson(delivery, Map.class))));

        JsonObject response = new JsonObject();
        response.addProperty("success", true);
        response.add("data", responseData);

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(response)));

        Map<String, Object> result = client.turboPartner().listWebhookDeliveries(ORG_ID, WEBHOOK_NAME, null, null);

        RecordedRequest request = server.takeRequest();
        assertEquals("GET", request.getMethod());
        assertEquals(
                "/partner/" + PARTNER_ID + "/orgs/" + ORG_ID + "/webhooks/" + WEBHOOK_NAME + "/deliveries",
                request.getPath()
        );

        assertTrue((Boolean) result.get("success"));
    }
}

File: packages/go-sdk/turbopartner_webhooks_test.go
package turbodocx

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// =============================================
// Webhook Management Tests
// =============================================

func TestCreateWebhook(t *testing.T) {
	t.Run("creates webhook successfully", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "POST", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks", r.URL.Path)
			assert.Equal(t, "Bearer TDXP-test-key", r.Header.Get("Authorization"))

			var body CreateWebhookRequest
			json.NewDecoder(r.Body).Decode(&body)
			assert.Equal(t, "my-signing-webhook", body.Name)
			assert.Equal(t, []string{"https://example.com/hook"}, body.URLs)
			assert.Equal(t, []string{"signature.document.completed"}, body.Events)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"id":     "webhook-uuid-789",
					"name":   "my-signing-webhook",
					"urls":   []string{"https://example.com/hook"},
					"events": []string{"signature.document.completed"},
					"secret": "whsec_abc123",
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.CreateWebhook(context.Background(), "org-uuid-456", &CreateWebhookRequest{
			Name:   "my-signing-webhook",
			URLs:   []string{"https://example.com/hook"},
			Events: []string{"signature.document.completed"},
		})

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.Equal(t, "my-signing-webhook", result.Data.Name)
		assert.NotEmpty(t, result.Data.Secret)
	})

	t.Run("returns auth error on 401", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusUnauthorized)
			json.NewEncoder(w).Encode(map[string]string{"error": "Unauthorized"})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		_, err := client.CreateWebhook(context.Background(), "org-uuid-456", &CreateWebhookRequest{
			Name:   "webhook",
			URLs:   []string{"https://example.com/hook"},
			Events: []string{"signature.document.completed"},
		})

		require.Error(t, err)
		authErr, ok := err.(*AuthenticationError)
		require.True(t, ok, "expected AuthenticationError")
		assert.NotNil(t, authErr)
	})
}

func TestListWebhooks(t *testing.T) {
	t.Run("lists webhooks with pagination", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks", r.URL.Path)
			assert.Equal(t, "10", r.URL.Query().Get("limit"))
			assert.Equal(t, "0", r.URL.Query().Get("offset"))

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"results":      []map[string]interface{}{{"id": "webhook-1", "name": "my-webhook"}},
					"totalRecords": 1,
					"limit":        10,
					"offset":       0,
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.ListWebhooks(context.Background(), "org-uuid-456", &ListWebhooksRequest{
			Limit:  IntPtr(10),
			Offset: IntPtr(0),
		})

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.Equal(t, 1, result.Data.TotalRecords)
		assert.Len(t, result.Data.Results, 1)
	})

	t.Run("lists webhooks without params", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"results": []interface{}{}, "totalRecords": 0, "limit": 50, "offset": 0,
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.ListWebhooks(context.Background(), "org-uuid-456", nil)

		require.NoError(t, err)
		assert.True(t, result.Success)
	})
}

func TestGetWebhook(t *testing.T) {
	t.Run("gets webhook by name", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks/my-signing-webhook", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data":    map[string]interface{}{"id": "webhook-uuid-789", "name": "my-signing-webhook"},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.GetWebhook(context.Background(), "org-uuid-456", "my-signing-webhook")

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.Equal(t, "my-signing-webhook", result.Data.Name)
	})

	t.Run("returns not found error on 404", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusNotFound)
			json.NewEncoder(w).Encode(map[string]string{"error": "Not Found"})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		_, err := client.GetWebhook(context.Background(), "org-uuid-456", "nonexistent")

		require.Error(t, err)
		notFoundErr, ok := err.(*NotFoundError)
		require.True(t, ok, "expected NotFoundError")
		assert.NotNil(t, notFoundErr)
	})
}

func TestUpdateWebhook(t *testing.T) {
	t.Run("updates webhook fields", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "PATCH", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks/my-signing-webhook", r.URL.Path)

			var body UpdateWebhookRequest
			json.NewDecoder(r.Body).Decode(&body)
			require.NotNil(t, body.IsActive)
			assert.False(t, *body.IsActive)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data":    map[string]interface{}{"id": "webhook-uuid-789", "name": "my-signing-webhook", "isActive": false},
			})
		}))
		defer server.Close()

		isActive := false
		client := newTestPartnerClient(t, server.URL)
		result, err := client.UpdateWebhook(context.Background(), "org-uuid-456", "my-signing-webhook", &UpdateWebhookRequest{
			IsActive: &isActive,
		})

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.False(t, result.Data.IsActive)
	})
}

func TestDeleteWebhook(t *testing.T) {
	t.Run("deletes webhook successfully", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "DELETE", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks/my-signing-webhook", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"message": "Webhook deleted",
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.DeleteWebhook(context.Background(), "org-uuid-456", "my-signing-webhook")

		require.NoError(t, err)
		assert.True(t, result.Success)
	})
}

func TestTestWebhook(t *testing.T) {
	t.Run("sends test event with no overrides", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "POST", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks/my-signing-webhook/test", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"deliveries": []interface{}{},
					"summary":    map[string]interface{}{"total": 1, "successful": 1, "failed": 0, "errors": []interface{}{}},
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.TestWebhook(context.Background(), "org-uuid-456", "my-signing-webhook", nil)

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.Equal(t, 1, result.Data.Summary.Successful)
	})

	t.Run("sends test event with event override", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			var body TestWebhookRequest
			json.NewDecoder(r.Body).Decode(&body)
			assert.Equal(t, "signature.document.voided", body.Event)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"deliveries": []interface{}{},
					"summary":    map[string]interface{}{"total": 1, "successful": 1, "failed": 0, "errors": []interface{}{}},
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.TestWebhook(context.Background(), "org-uuid-456", "my-signing-webhook", &TestWebhookRequest{
			Event: "signature.document.voided",
		})

		require.NoError(t, err)
		assert.True(t, result.Success)
	})
}

func TestListWebhookDeliveries(t *testing.T) {
	t.Run("lists deliveries with pagination", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/partner/test-partner-id/orgs/org-uuid-456/webhooks/my-signing-webhook/deliveries", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success": true,
				"data": map[string]interface{}{
					"results": []map[string]interface{}{
						{"id": "del-1", "event": "signature.document.completed", "statusCode": 200},
					},
					"totalRecords": 1,
					"limit":        50,
					"offset":       0,
				},
			})
		}))
		defer server.Close()

		client := newTestPartnerClient(t, server.URL)
		result, err := client.ListWebhookDeliveries(context.Background(), "org-uuid-456", "my-signing-webhook", nil)

		require.NoError(t, err)
		assert.True(t, result.Success)
		assert.Equal(t, 1, result.Data.TotalRecords)
		assert.Len(t, result.Data.Results, 1)
	})
}
